### PR TITLE
Cal/configurable engine

### PR DESCRIPTION
**Pull request by McCallisterRomer**
* Refactored Engine into smaller pieces.
* Replaced the System/Impl convention with modules which can be stubbed out (e.g. disabling physics in NcEditor or running in headless mode).
* Added concept of a primary execution graph. Modules can build workloads (jobs or task graphs) that automatically attach to execution stages, allowing us to run most of the "game loop" by executing the primary task graph.

### CODE REVIEW

**Comment by ClickerMonkey** [comment]
There's barely any code here, you call this a pull request?

:-)

File: nc/source/graphics/GraphicsModuleImpl.cpp [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+#include "window/WindowImpl.h"
+
+#include "optick/optick.h"
+
+namespace nc::graphics
+{
+    auto BuildGraphicsModule(bool enableModule, Registry* reg, window::WindowImpl* window) -> std::unique_ptr<GraphicsModule>
+    {
+        /** @todo allow stub */
+        (void)enableModule;
```
Comment (McCallisterRomer) [inline nc/source/graphics/GraphicsModuleImpl.cpp]: oops... this should probably be added now.

File: nc/include/NcEngine.h [hunk 2]
```diff
@@ -1,37 +1,96 @@
...
+    /** @brief Flags controlling initialization of engine modules. */
+    enum class EngineInitFlags : uint8_t
+    {
+        /** Standard initialization. */
+        None = 0,
+
+        /** Disable the graphics module. */
+        Headless = 1,
+
+        /** Disables the physics module. */
```
Comment (McCallisterRomer) [inline nc/include/NcEngine.h]: Disable

File: nc/include/NcEngine.h [hunk 3]
```diff
@@ -1,37 +1,96 @@
...
 
-    auto InitializeNcEngine(std::string_view configPath, bool useEditorMode = false) -> std::unique_ptr<NcEngine>;
+    /** @brief Flags controlling initialization of engine modules. */
+    enum class EngineInitFlags : uint8_t
+    {
+        /** Standard initialization. */
+        None = 0,
+
+        /** Disable the graphics module. */
+        Headless = 1,
```
Comment (McCallisterRomer) [inline nc/include/NcEngine.h]: Is it better to be consistent and use "NoGraphics" here?

File: nc/source/audio/AudioModuleImpl.cpp [hunk 4]
```diff
@@ -22,14 +23,20 @@ namespace
...
         return system->WriteToDeviceBuffer(static_cast<double*>(outputBuffer));
     }
 }
 
 namespace nc::audio
 {
-    AudioSystemImpl::AudioSystemImpl(Registry* registry)
+    auto BuildAudioModule(Registry* reg) -> std::unique_ptr<AudioModule>
+    {
+        /** @note If we make a stub version of this, consider the cast in AudioSystemCallback. */
```
Comment (McCallisterRomer) [inline nc/source/audio/AudioModuleImpl.cpp]: Make this a todo with a github issue. We may need to stub this for testing on github runners.

File: nc/include/Module.h [hunk 5]
```diff
@@ -0,0 +1,28 @@
+#pragma once
+
+#include <vector>
+
+namespace nc
+{
+    struct Job;
+
+    /** @brief Base class for modules that attach worloads to the core execution graph. */
```
Comment (jaremieromer) [inline nc/include/Module.h]: *workloads

File: nc/include/NcEngine.h (hunk 3 above)
Comment (jaremieromer) [reply nc/include/NcEngine.h]: Yes - my preference is for "NoGraphics"

Comment (jaremieromer) [reply nc/include/NcEngine.h]: It makes it %100 clear what is being omitted.

File: nc/include/audio/AudioModule.h [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+#include "ecs/Entity.h"
+#include "Module.h"
+
+namespace nc
+{
+    /** @brief Interface for general audio functionality. */
+    struct AudioModule : public Module
+    {
+        /**
+         * @brief Regester an object to be the reference point audio.
```
Comment (jaremieromer) [inline nc/include/audio/AudioModule.h]: *Register

File: nc/include/graphics/GraphicsModule.h [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+         * @brief Get the main Camera.
+         * @return Camera*
+         */
+        virtual auto GetCamera() noexcept -> Camera* = 0;
+
+        /**
+         * @brief Set a custom ui to receive draw callbacks during rendering.
+         * @param ui A pointer to a valid IUI implementation.
+         */
+        virtual void SetUi(ui::IUI* ui) noexcept = 0;
```
Comment (jaremieromer) [inline nc/include/graphics/GraphicsModule.h:34]: Does graphics work right now if you don't set this? I should test that

File: nc/source/ecs/ParticleEmitterSystem.h [hunk 8]
```diff
@@ -4,20 +4,27 @@
 #include "ecs/component/ParticleEmitter.h"
 #include "particle/EmitterState.h"
 
-namespace nc::graphics { class Graphics; }
-
 namespace nc::ecs
 {
+    /** @todo I skipped making this a module because Jare was working
```
Comment (jaremieromer) [inline nc/source/ecs/ParticleEmitterSystem.h:9]: Thanks

File: nc/source/engine/Modules.h [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+#include "audio/AudioModule.h"
+#include "ecs/LogicModule.h"
+#include "ecs/ParticleEmitterSystem.h"
+#include "graphics/GraphicsModule.h"
+#include "physics/PhysicsModule.h"
+#include "scene/SceneModule.h"
+
+namespace nc
+{
+    struct Modules
```
Comment (jaremieromer) [inline nc/source/engine/Modules.h:12]: SO NICE

File: nc/source/engine/Runtime.cpp [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+#include "scene/SceneModuleImpl.h"
+
+namespace
+{
+    auto BuildContext() -> nc::Context
+    {
+        V_LOG("BuildContext()");
+        return nc::Context
+        {
+            .registry = nc::Registry{nc::config::GetMemorySettings().maxTransforms},
```
Comment (jaremieromer) [inline nc/source/engine/Runtime.cpp:17]: This is exactly what we were planning, nice job!

File: nc/source/graphics/GraphicsModuleImpl.cpp [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+
+    auto GraphicsModuleImpl::BuildWorkload() -> std::vector<Job>
+    {
+        return std::vector<Job>
+        {
+            Job{ [this]{Run();}, "GraphicsModule", HookPoint::Render }
+        };
+    }
+
+    void GraphicsModuleImpl::Run()
```
Comment (jaremieromer) [inline nc/source/graphics/GraphicsModuleImpl.cpp:116]: I can't believe you figured this all out for graphics

**Review by jaremieromer — APPROVED** [review]
I'm so excited - this is great.

File: nc/include/graphics/GraphicsModule.h (hunk 7 above)
Comment (McCallisterRomer) [reply nc/include/graphics/GraphicsModule.h:34]: Yes

**Pull request merged**